### ISSUE TITLE
Type error not detected by Mypy in flatten implementation

### ISSUE THREAD

**Issue by rs86** [python/mypy#1004]
I was playing around with Mypy and I thought I'd implement flatten with type checks.
```
from typing import List               
                                      
def flatten(lst: List) -> List:       
    if len(lst) == 1:                 
        return lst                    
    if len(lst) > 1:                  
        h, *t = lst                   
        return flatten(h) + flatten(t)
    return lst                        
                                      
print(flatten([1,2,[3,4]]))           
```
Python will raise a TypeError:
```
Type:  [1, 2, [3, 4]]                                                                │
Type:  1                                                                             │
Traceback (most recent call last):                                                   │
  File "typecheck.py", line 13, in <module>                                          │
    print(flatten([1,2,[3,4]]))                                                      │
  File "typecheck.py", line 10, in flatten                                           │
    return flatten(h) + flatten(t)                                                   │
  File "typecheck.py", line 6, in flatten                                            │
    if len(lst) == 1:                                                                │
TypeError: object of type 'int' has no len()
```
But Mypy doesn't warn me about it.

Is this a bug? Should it have inferred that h, *t will destructure a list and h will be possibly a non-list value?

tks!

**Comment by kirbyfan64** [comment]
The problem is that you just said `List`, which is the same thing as `List[Any]`, so Mypy doesn't know what exactly is inside the list. I think you could do something like:

```python
T = TypeVar('T')
def flatten(lst: List[T]) -> List:
    ...
```

**Comment by JukkaL** [comment]
Also, PEP 484 syntax probably isn't expressive enough to give a precise signature for `flatten`.

**Comment by gvanrossum** [comment]
Closing, as the use of plain `List` means `List[Any]` and then mypy intentionally shuts up, trusting you and/or the runtime.
